pkg/operator/ceph/reporting/status.go

```diff
@@ -0,0 +1,82 @@
+/*
+Copyright 2020 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reporting
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/coreos/pkg/capnslog"
+	"github.com/pkg/errors"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var logger = capnslog.NewPackageLogger("github.com/rook/rook", "ceph-operator-status")
+
+// UpdateStatus updates an object with a given status. The object is updated with the latest version
+// from the server on a successful update.
+func UpdateStatus(client client.Client, obj client.Object) error {
+	nsName := types.NamespacedName{
+		Namespace: obj.GetNamespace(),
+		Name:      obj.GetName(),
+	}
+	baseErrMsg := fmt.Sprintf("failed to update object %q status", nsName.String())
+
+	// Try to update the status
+	err := client.Status().Update(context.Background(), obj)
+	// If the object doesn't exist yet, we need to initialize it
+	if kerrors.IsNotFound(err) {
+		err = client.Update(context.Background(), obj)
+	}
+	if err != nil {
+		// TODO: ask Seb why the code retries; there is no point if there is a conflict or invalid
+		if !kerrors.IsConflict(err) && !kerrors.IsInvalid(err) {
```

Comment:
When looking at log files I frequently see errors about the status not being updated because of a conflict. It really seems like we should be able to try updating the status multiple times until it succeeds, instead of just failing. Although that would involve first retrieving a new object, so it's not a very simple issue to solve. When testing all these changes please do look for messages like this and consider how common it is. It's quite random though, I can't grasp how frequent it is.

Reply:
Right. I agree, but this retry will not succeed on the second try if the first one fails no matter what because if there is a conflict, we need to re-do the status update code before the second try anyway to resync and avoid the conflict. This seems redundant to me.

Reply:
On the topic of retrying, I think we should do that, and I think we might want to treat failures to update status as failures that require requeuing updates, because statuses are an important part of the user experience.

I think one of the reasons we get so many conflicts is be cause we have the main reconcile loops running simultaneously with health checkers that don't have mutex locks. We could avoid a lot of the issue by having a mutex lock for each health checker that would prevent both from reading, modifying, and updating the resource at the same time.

Reply:
Update: I implemented retrying for the object store status, and it fixed the issue I experienced in my testing.